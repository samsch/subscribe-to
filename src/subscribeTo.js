import React from 'react';

const subscribeTo = getSubscribable => Component => {
  class SubscribedTo extends React.Component {
    constructor(props) {
      super(props);
      this.onModelUpdate = () => {
        this.forceUpdate();
      };
    }
    componentDidMount() {
      this.subscribables = [].concat(getSubscribable(this.props));
      this.subscribables.forEach(subscribable => {
        subscribable.subscribe(this.onModelUpdate);
      });
    }
    componentWillUnmount() {
      this.subscribables.forEach(subscribable => {
        subscribable.unsubscribe(this.onModelUpdate);
      });
    }
    render() {
      return <Component {...this.props} />;
    }
  }
  return SubscribedTo;
};

export default subscribeTo;
