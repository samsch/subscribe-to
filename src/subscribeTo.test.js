/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import subscribeTo from './subscribeTo';

describe('subscribeTo', () => {
  test('subscribes component to updates from given store', () => {
    const jf = jest.fn();
    const Component = props => {
      jf(props.store.data);
      return <span>Data: {props.store.data}</span>;
    };

    const SubscribedComponent = subscribeTo(props => props.store)(Component);

    const store = (() => {
      let listener;
      const store = {
        data: 0,
      };
      store.subscribe = callback => {
        listener = callback;
      };
      store.unsubscribe = () => {
        listener = null;
      };
      store.update = () => {
        store.data++;
        listener();
      };
      return store;
    })();

    const component = renderer.create(<SubscribedComponent store={store} />);
    expect(component.toJSON()).toMatchSnapshot();
    expect(jf).toHaveBeenCalledWith(0);
    store.update();
    expect(component.toJSON()).toMatchSnapshot();
    expect(jf).toHaveBeenCalledWith(1);
  });
});
