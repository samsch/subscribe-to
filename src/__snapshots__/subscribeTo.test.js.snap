// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subscribeTo subscribes component to updates from given store 1`] = `
<span>
  Data: 
  0
</span>
`;

exports[`subscribeTo subscribes component to updates from given store 2`] = `
<span>
  Data: 
  1
</span>
`;
